### ISSUE TITLE
Sync Batch Norm for Pytorch

### ISSUE THREAD

**Issue by msalvaris** [horovod/horovod#1384]
There are sync batch norm implementations for PyTorch distributed and Apex but can't seem to find one for Horovod. Is there a technical reason why there isn't one?

**Comment by abditag2** [comment]
@msalvaris sync batch norm is not implemented for horovod yet but it is in our todo list coming out soon.

**Comment by msalvaris** [comment]
@abditag2 That is great to hear. Would the Horovod implementation of sync batch norm simply be like the PyTorch one except replacing the distributed methods such as all_reduce with Horovod ones or is there some further technical issues that need to be overcome?

**Comment by abditag2** [comment]
Most likely. It will also be implemented for keras and Tensorflow.

**Comment by atranitell** [comment]
Hello, I have implemented a sync-bn based on pytorch and horovod.
See: https://github.com/atranitell/Synchronized-BatchNorm-PyTorch-Horovod/blob/master/sync_bn.py

```bash
[1,0]<stdout>:affine: True track_running_stats False
[1,0]<stdout>:[INFERENCE PHASE-1]
[1,0]<stdout>:[NAME]: model.outputs.0-108.2299  [SHAPE]: [8, 1000]       [SUM]: 190.1940002, [MEAN]: 0.0237742, [MAX]: 2.6216030, [MIN]: -2.8100750, [VAR]: 0.5494183
[1,0]<stdout>:[NAME]: model.outputs.1-56.0818   [SHAPE]: [8, 1000]       [SUM]: 165.8392944, [MEAN]: 0.0207299, [MAX]: 2.7186546, [MIN]: -3.5223341, [VAR]: 0.5672572
[1,0]<stdout>:[NAME]: model.outputs.2-62.6330   [SHAPE]: [8, 1000]       [SUM]: 181.9670258, [MEAN]: 0.0227459, [MAX]: 2.6616516, [MIN]: -2.7214265, [VAR]: 0.5471250
[1,0]<stdout>:[NAME]: sync_model.outputs.0-78.4819 [SHAPE]: [8, 1000]       [SUM]: 190.1936188, [MEAN]: 0.0237742, [MAX]: 2.6216090, [MIN]: -2.8100815, [VAR]: 0.5494181
[1,0]<stdout>:[NAME]: sync_model.outputs.1-56.4089 [SHAPE]: [8, 1000]       [SUM]: 165.8392487, [MEAN]: 0.0207299, [MAX]: 2.7186563, [MIN]: -3.5223362, [VAR]: 0.5672570
[1,0]<stdout>:[NAME]: sync_model.outputs.2-51.5120 [SHAPE]: [8, 1000]       [SUM]: 181.9673615, [MEAN]: 0.0227459, [MAX]: 2.6616564, [MIN]: -2.7214241, [VAR]: 0.5471250
[1,0]<stdout>:[TRAINING PHASE]
[1,0]<stdout>:[NAME]: model.outputs.0-55.6757   [SHAPE]: [8, 1000]       [SUM]: 190.1940002, [MEAN]: 0.0237742, [MAX]: 2.6216030, [MIN]: -2.8100750, [VAR]: 0.5494183
[1,0]<stdout>:[NAME]: model.outputs.1-68.0540   [SHAPE]: [8, 1000]       [SUM]: -117.8934021, [MEAN]: -0.0147367, [MAX]: 2.8053782, [MIN]: -3.3515174, [VAR]: 0.5563220
[1,0]<stdout>:[NAME]: model.outputs.2-122.4730  [SHAPE]: [8, 1000]       [SUM]: -394.6956787, [MEAN]: -0.0493370, [MAX]: 2.5790358, [MIN]: -2.8156824, [VAR]: 0.5413557
[1,0]<stdout>:[NAME]: sync_model.outputs.0-166.4059 [SHAPE]: [8, 1000]       [SUM]: 190.1939545, [MEAN]: 0.0237742, [MAX]: 2.6216037, [MIN]: -2.8100758, [VAR]: 0.5494182
[1,0]<stdout>:[NAME]: sync_model.outputs.1-240.0510 [SHAPE]: [8, 1000]       [SUM]: -117.8935699, [MEAN]: -0.0147367, [MAX]: 2.8053713, [MIN]: -3.3515224, [VAR]: 0.5563222
[1,0]<stdout>:[NAME]: sync_model.outputs.2-247.0720 [SHAPE]: [8, 1000]       [SUM]: -394.6961365, [MEAN]: -0.0493370, [MAX]: 2.5790341, [MIN]: -2.8156407, [VAR]: 0.5413551
[1,0]<stdout>:[INFERENCE PHASE-2]
[1,0]<stdout>:[NAME]: model.outputs.0-92.9980   [SHAPE]: [8, 1000]       [SUM]: -838.2194214, [MEAN]: -0.1047774, [MAX]: 2.7204847, [MIN]: -2.9403963, [VAR]: 0.5573729
[1,0]<stdout>:[NAME]: model.outputs.1-74.8370   [SHAPE]: [8, 1000]       [SUM]: -830.7557373, [MEAN]: -0.1038445, [MAX]: 2.6922336, [MIN]: -3.0616376, [VAR]: 0.5771688
[1,0]<stdout>:[NAME]: model.outputs.2-79.0393   [SHAPE]: [8, 1000]       [SUM]: -798.6572266, [MEAN]: -0.0998322, [MAX]: 3.2130013, [MIN]: -2.9527130, [VAR]: 0.5481955
[1,0]<stdout>:[NAME]: sync_model.outputs.0-59.4788 [SHAPE]: [8, 1000]       [SUM]: -838.2209473, [MEAN]: -0.1047776, [MAX]: 2.7205055, [MIN]: -2.9404104, [VAR]: 0.5573736
[1,0]<stdout>:[NAME]: sync_model.outputs.1-58.3360 [SHAPE]: [8, 1000]       [SUM]: -830.7567749, [MEAN]: -0.1038446, [MAX]: 2.6922214, [MIN]: -3.0616069, [VAR]: 0.5771680
[1,0]<stdout>:[NAME]: sync_model.outputs.2-84.6570 [SHAPE]: [8, 1000]       [SUM]: -798.6591797, [MEAN]: -0.0998324, [MAX]: 3.2129552, [MIN]: -2.9527016, [VAR]: 0.5481985
[1,0]<stdout>:
[1,0]<stdout>:
[1,0]<stdout>:affine: True track_running_stats True
[1,0]<stdout>:[INFERENCE PHASE-1]
[1,0]<stdout>:[NAME]: model.outputs.0-43.8871   [SHAPE]: [8, 1000]       [SUM]: 73.2726212, [MEAN]: 0.0091591, [MAX]: 0.5324938, [MIN]: -0.5004993, [VAR]: 0.0269264
[1,0]<stdout>:[NAME]: model.outputs.1-29.7771   [SHAPE]: [8, 1000]       [SUM]: 79.4624710, [MEAN]: 0.0099328, [MAX]: 0.5239193, [MIN]: -0.5046086, [VAR]: 0.0266207
[1,0]<stdout>:[NAME]: model.outputs.2-35.8410   [SHAPE]: [8, 1000]       [SUM]: 77.2687073, [MEAN]: 0.0096586, [MAX]: 0.5608573, [MIN]: -0.4926656, [VAR]: 0.0273252
[1,0]<stdout>:[NAME]: sync_model.outputs.0-38.7599 [SHAPE]: [8, 1000]       [SUM]: 73.2726212, [MEAN]: 0.0091591, [MAX]: 0.5324938, [MIN]: -0.5004993, [VAR]: 0.0269264
[1,0]<stdout>:[NAME]: sync_model.outputs.1-42.7032 [SHAPE]: [8, 1000]       [SUM]: 79.4624710, [MEAN]: 0.0099328, [MAX]: 0.5239193, [MIN]: -0.5046086, [VAR]: 0.0266207
[1,0]<stdout>:[NAME]: sync_model.outputs.2-48.8539 [SHAPE]: [8, 1000]       [SUM]: 77.2687073, [MEAN]: 0.0096586, [MAX]: 0.5608573, [MIN]: -0.4926656, [VAR]: 0.0273252
[1,0]<stdout>:[TRAINING PHASE]
[1,0]<stdout>:[NAME]: model.outputs.0-57.8671   [SHAPE]: [8, 1000]       [SUM]: 190.1940002, [MEAN]: 0.0237742, [MAX]: 2.6216030, [MIN]: -2.8100750, [VAR]: 0.5494183
[1,0]<stdout>:[NAME]: model.outputs.1-112.4310  [SHAPE]: [8, 1000]       [SUM]: -117.8934021, [MEAN]: -0.0147367, [MAX]: 2.8053782, [MIN]: -3.3515174, [VAR]: 0.5563220
[1,0]<stdout>:[NAME]: model.outputs.2-97.1479   [SHAPE]: [8, 1000]       [SUM]: -394.6956787, [MEAN]: -0.0493370, [MAX]: 2.5790358, [MIN]: -2.8156824, [VAR]: 0.5413557
[1,0]<stdout>:[NAME]: sync_model.outputs.0-203.5007 [SHAPE]: [8, 1000]       [SUM]: 190.1939545, [MEAN]: 0.0237742, [MAX]: 2.6216037, [MIN]: -2.8100758, [VAR]: 0.5494182
[1,0]<stdout>:[NAME]: sync_model.outputs.1-256.1831 [SHAPE]: [8, 1000]       [SUM]: -117.8931427, [MEAN]: -0.0147366, [MAX]: 2.8053741, [MIN]: -3.3515232, [VAR]: 0.5563222
[1,0]<stdout>:[NAME]: sync_model.outputs.2-277.2660 [SHAPE]: [8, 1000]       [SUM]: -394.6949768, [MEAN]: -0.0493369, [MAX]: 2.5790620, [MIN]: -2.8156748, [VAR]: 0.5413552
[1,0]<stdout>:[INFERENCE PHASE-2]
[1,0]<stdout>:[NAME]: model.outputs.0-50.8077   [SHAPE]: [8, 1000]       [SUM]: -82.1022491, [MEAN]: -0.0102628, [MAX]: 0.2733500, [MIN]: -0.3195015, [VAR]: 0.0071264
[1,0]<stdout>:[NAME]: model.outputs.1-66.3068   [SHAPE]: [8, 1000]       [SUM]: -79.3727341, [MEAN]: -0.0099216, [MAX]: 0.2692815, [MIN]: -0.2908769, [VAR]: 0.0072672
[1,0]<stdout>:[NAME]: model.outputs.2-48.0020   [SHAPE]: [8, 1000]       [SUM]: -81.5532227, [MEAN]: -0.0101942, [MAX]: 0.2574194, [MIN]: -0.3002077, [VAR]: 0.0071647
[1,0]<stdout>:[NAME]: sync_model.outputs.0-67.1532 [SHAPE]: [8, 1000]       [SUM]: -81.9084091, [MEAN]: -0.0102386, [MAX]: 0.2692602, [MIN]: -0.3135209, [VAR]: 0.0068271
[1,0]<stdout>:[NAME]: sync_model.outputs.1-46.2432 [SHAPE]: [8, 1000]       [SUM]: -79.1543961, [MEAN]: -0.0098943, [MAX]: 0.2645385, [MIN]: -0.2849320, [VAR]: 0.0069433
[1,0]<stdout>:[NAME]: sync_model.outputs.2-35.9240 [SHAPE]: [8, 1000]       [SUM]: -81.3666458, [MEAN]: -0.0101708, [MAX]: 0.2528106, [MIN]: -0.2988353, [VAR]: 0.0068605
[1,0]<stdout>:
[1,0]<stdout>:
```

There is a slight error (about 1e-6), and I'm not clear currently.

KJ,
Best

**Comment by abditag2** [comment]
@atranitell Do you want to make a PR?

**Comment by atranitell** [comment]
@abditag2 Emmm.. Surely, I am glad to do that. But, should I place the "sync_bn" codes at "/examples"?
In addition, I just implement a pytorch version.

Best,
KJ

**Comment by weberxie** [comment]
@atranitell Nice work! Any updates on this feature? Could you please update the it with latest version?

**Comment by atranitell** [comment]
@weberxie probably, I will update the latest version and test it at this weekend, compared with the torch.dist.

**Comment by alsrgv** [comment]
@atranitell, I haven't seen your code until now - do you want to collaborate on #1923?  It's currently using PyTorch's internal kernels and would be great to switch to an implementation that does not use internals.  One requirement for it is to be able to hot swap BatchNorm to SyncBatchNorm in models like torchvision.resnet.
